src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs

```diff
@@ -187,53 +236,93 @@ public async Task<TrialResult> RunAsync(CancellationToken ct = default)
+            var logger = serviceProvider.GetRequiredService<IChannel>();
             var trialNum = 0;
             var tuner = serviceProvider.GetService<ITuner>();
             Contracts.Assert(tuner != null, "tuner can't be null");
-            while (true)
+
+            while (!_globalCancellationTokenSource.Token.IsCancellationRequested)
             {
-                if (_cts.Token.IsCancellationRequested)
-                {
-                    break;
-                }
                 var setting = new TrialSettings()
                 {
                     TrialId = trialNum++,
                     Parameter = Parameter.CreateNestedParameter(),
                 };
-
                 var parameter = tuner.Propose(setting);
                 setting.Parameter = parameter;
-                monitor?.ReportRunningTrial(setting);
-                var runner = serviceProvider.GetRequiredService<ITrialRunner>();
 
+                monitor?.ReportRunningTrial(setting);
                 try
                 {
-                    var trialResult = runner.Run(setting);
-                    monitor?.ReportCompletedTrial(trialResult);
-                    tuner.Update(trialResult);
-
-                    var error = _settings.IsMaximize ? 1 - trialResult.Metric : trialResult.Metric;
-                    if (error < _bestError)
+                    using (var trialCancellationTokenSource = new CancellationTokenSource())
+                    using (var deregisterCallback = _globalCancellationTokenSource.Token.Register(() =>
                     {
-                        _bestTrialResult = trialResult;
-                        _bestError = error;
-                        monitor?.ReportBestTrial(trialResult);
+                        // only force-canceling running trials when there's completed trials.
+                        // otherwise, wait for the current running trial to be completed.
+                        if (_bestTrialResult != null)
+                            trialCancellationTokenSource.Cancel();
+                    }))
+                    using (var performanceMonitor = serviceProvider.GetService<IPerformanceMonitor>())
+                    using (var runner = serviceProvider.GetRequiredService<ITrialRunner>())
+                    {
+                        performanceMonitor.MemoryUsageInMegaByte += (o, m) =>
+                        {
+                            if (_settings.MaximumMemoryUsageInMegaByte is double d && m > d && !trialCancellationTokenSource.IsCancellationRequested)
+                            {
+                                logger.Trace($"cancel current trial {setting.TrialId} because it uses {m} mb memory and the maximum memory usage is {d}");
+                                trialCancellationTokenSource.Cancel();
+
+                                GC.AddMemoryPressure(Convert.ToInt64(m) * 1024 * 1024);
+                                GC.Collect();
```

Comment:
why we are doing that?

Reply:
You mean `GC.Collect` or `GC.AddMemoryPressure` part?

Reply:
yes

Reply:
This is because if we hit that code, it means the current trial uses memory that exceed the limit. Therefore we would like to run a GC to collect memory after that trial get cancelled.

`GC.AddMemoryPressure` is mostly for LightGbm which might allocate large chunk of unmanaged memory.

Reply:
### Resolved

Reply:
Wouldn't GC kick off automatically when it needs more memory? at that time will collect any stale objects. I am asking when you are forcing that to happen at this time?

Reply:
But it wouldn't if GC thinks it still has enough memory. And in that case, if we didn't run GC, the next trial will likely to be canceled even if it's not using a lot of memory because the memory used by the last trial is not released.

Reply:
> But it wouldn't if GC thinks it still has enough memory.

Could you please talk more about that? Why will GC think there will not be enough memory? I expect GC will run to collect any unneeded objects at that time which should free a lot of memory. no? I am not objecting here, I am trying to understand the case.

Reply:
> I expect GC will run to collect any unneeded objects at that time which should free a lot of memory. no?

Only if GC thinks there's not enough memory and runs a collection, but that's not always the case when a trial gets canceled. For example, if the user sets a very small number for maximum memory usage.

Reply:
I admit I am not fully understanding the whole picture :-) I'll leave it to you if you are sure about that. You may ignore my comment then.